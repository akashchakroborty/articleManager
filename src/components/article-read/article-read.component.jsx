import React, { useState, useEffect } from "react";
import ArticlePannel from "../article-pannel/article-pannel.component";
import ArticleImagePannel from "../article-image-pannel/article-image-pannel.component";
import {
  ArticleReadContainer,
  ContentContainer,
  Content,
  SidePannel,
  ClapCount,
  ClapWrapper,
} from "./article-read.styles";
import {
  ActionContainer,
  ActionImage,
  ArticleBoby,
  TitleContainer,
  DescriptionContainer,
} from "../article-pannel/article-pannel.styles";
import { connect } from "react-redux";
import { incrementClapsInArticle } from "../../redux/articles/articles.action";
import clapIcon from "../../assets/clapIcon.png";

export const ArticleRead = ({ article, incrementClapsInArticle }) => {
  const { id, img, content, ...restprops } = article;
  const { title, description } = article;
  let { claps } = article;
  const splitContent = content.split(/\r?\n/);

  const clapIncrementHandeller = () => {
    incrementClapsInArticle(id, { ...article, claps: ++claps });
  };
  const [scroll, setScroll] = useState({
    hide: true,
    scrollPos: -212,
  });
  const { hide, scrollPos } = scroll;

  const isBottom = (el) => {
    return el.getBoundingClientRect().bottom <= window.innerHeight;
  };

  const handleScroll = () => {
    const wrappedElement = document.getElementById("bottomClap");
    setScroll({
      ...scroll,
      scrollPos: document.body.getBoundingClientRect().top,
      hide:
        document.body.getBoundingClientRect().top > scrollPos ||
        isBottom(wrappedElement),
    });
  };

  React.useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return function () {
      window.removeEventListener("scroll", handleScroll);
    };
  }, []);

  return (
    <ArticleReadContainer>
      <ArticlePannel {...restprops} isArticleRead />
      <ArticleImagePannel img={img} isArticleRead id="article-read-image" />
      <ContentContainer className="article-content">
        {splitContent.map((content, index) => (
          <Content key={index}>{content}</Content>
        ))}
      </ContentContainer>
      {!hide && (
        <SidePannel className="article-side-pannel">
          <ArticleBoby>
            <TitleContainer>{title}</TitleContainer>
            <DescriptionContainer>{description}</DescriptionContainer>
          </ArticleBoby>
          <ClapWrapper>
            <ActionContainer
              className="clap-icon"
              onClick={clapIncrementHandeller}
            >
              <ActionImage actionUrl={clapIcon} />
            </ActionContainer>
            <ClapCount className="clap-count">{claps}</ClapCount>
          </ClapWrapper>
        </SidePannel>
      )}
      <ClapWrapper id="bottomClap">
        <ActionContainer className="clap-icon" onClick={clapIncrementHandeller}>
          <ActionImage actionUrl={clapIcon} />
        </ActionContainer>
        <ClapCount className="clap-count">{claps} Claps</ClapCount>
      </ClapWrapper>
    </ArticleReadContainer>
  );
};

export default connect(null, { incrementClapsInArticle })(ArticleRead);
