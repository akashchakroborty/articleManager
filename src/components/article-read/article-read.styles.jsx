import styled from "styled-components";

export const ArticleReadContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  @media screen and (max-width: 800px) {
    width: 80%;
  }
`;

ArticleReadContainer.displayName = "ArticleReadContainer";

export const ContentContainer = styled.div`
  display: flex;
  flex-direction: column;
  text-align: justify;
  letter-spacing: -0.022em;
  line-height: 1.12;
  align-items: center;
  margin-bottom: 30px;
  word-break: break-all;
  > *:first-child {
    margin-top: 1.95em;
    text-align: left;
    margin-bottom: -0.28em;
    font-size: 34px;
    font-weight: 600;
    @media screen and (max-width: 800px) {
      margin-top: 0px;
      word-break: break-word;
    }
  }
`;

export const Content = styled.div`
  line-height: 32px;
  color: rgba(41, 41, 41, 1);
  margin-top: 0.86em;
  font-size: 21px;
  margin-bottom: -0.46em;
  letter-spacing: -0.003em;
`;

Content.displayName = "Content";

export const SidePannel = styled.div`
  position: fixed;
  width: 160px;
  top: 40%;
  left: 20px;
  text-align: left;
  @media screen and (max-width: 800px) {
    display: none;
  }
`;

SidePannel.displayName = "SidePannel";

export const ClapWrapper = styled.div`
  display: flex;
`;

ClapWrapper.displayName = "ClapWrapper";

export const ClapCount = styled.div`
  align-self: flex-end;
  margin: 10px;
  color: grey;
`;
