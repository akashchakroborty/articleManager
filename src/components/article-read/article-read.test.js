import React from "react";
import { shallow } from "enzyme";
import { ArticleRead } from "./article-read.component";

describe("ArticleRead component", () => {
  let wrapper;
  let mockIncrementClapsInArticle;
  let useEffect;
  beforeEach(() => {
    mockIncrementClapsInArticle = jest.fn();
    useEffect = jest.spyOn(React, "useEffect").mockImplementation((f) => f());
    const mockProps = {
      article: {
        title: "Mock data",
        description: "Mock data",
        author: "Mock data",
        content: "Mock data",
        id: "Mock data",
        img: "Mock data",
        avatar: "Mock data",
        claps: 38,
      },
      incrementClapsInArticle: mockIncrementClapsInArticle,
    };

    wrapper = shallow(<ArticleRead {...mockProps} />);
  });

  it("should render ArticleRead component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should increment claps when clap Icon is clicked", () => {
    wrapper.find("ActionContainer").simulate("click");
    expect(mockIncrementClapsInArticle).toHaveBeenCalled();
  });
});
