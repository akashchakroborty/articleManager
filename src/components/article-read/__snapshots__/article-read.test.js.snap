// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleRead component should render ArticleRead component 1`] = `ShallowWrapper {}`;
